designs/drift-hash-versioning.md

```diff
@@ -0,0 +1,180 @@
+
+As a result, two annotations will be used analyze and orchestrate drift: 
+
+* nodepool-hash: (hash of the nodeclaim template on the nodepool)
+* nodepool-hash-version: (the NodePool CRD version defined by Karpenter)
+
+As part of option 1, Karpenter will need to expanded its `clusterrole` to have read permissions for the CRDs. The `clusterrole` permission will be scoped down to being on the `NodePool`. This will look as such:
+
+```
+...
+- apiGroups: ["apiextensions.k8s.io"]
+  resources: ["customresourcedefinitions"]
+  verbs: ["get", "watch"]
+  resourceNames: ["nodepools.karpenter.sh"]
+...
+```
+
+### Option 1A(Recommended): Fallback to rehashing on the NodeClaim 
+
+Option 1 will have two case for drift:
+
+1. When the CRD `nodepool-hash-version` matches between the `NodePool` and `NodeClaim`:
+    1. Drift will work as normal and the `NodePool` hash will be compared to evaluate for drift 
+2. When the CRD `nodepool-hash-version` does not match between the `NodePool` and `NodeClaim`:
+    1. Karpenter would recalculate the hash of the `Nodepool` and back-propagate the annotation to the NodeClaims
+      - Karpenter will need not back-propagate the `nodepool-crd-hash` annotation `NodeClaims` that are considered drifted.
+    2. Drift will work as normal once a `NodePool` field is updated
+
+Pros/Cons
+
+- 👍👍 Simple implementation effort 
+- 👍 Cover general cases of updates to the `NodePool` CRD or the hashing mechanism
+- 👎 Nodes that were considered drifted before the Karpenter upgrade are not able to be un-drifted after a Karpenter upgrade 
+- 👎 Does not handle cases where users change `NodePool` resources as Karpenter is updating
+
+### Option 1B: Fallback to individual field evaluation for drift
+
+Similar to Option 1, there will be two cases for drift:
+
+1. When the `nodepool-hash-version` matches between the `NodePool` and `NodeClaim`:
+    1. Drift will work as normal and the `NodePool` hash will be compared to evaluate for drift 
+2. When the `nodepool-hash-version` does not match between the `NodePool` and `NodeClaim`:
+    1. Karpenter will individually check each field for drift. *This would be an adaptation of the (Deployment Controller)[https://github.com/kubernetes/kubernetes/blob/83e6636096548f4b49a94ece6bc4a26fd67e4449/pkg/controller/deployment/util/deployment_util.go#L603] for podTemplateSpec relationship to pods.*
+    2. Karpenter will need to populate the `NodeClaim` with values that were used in launching the instances. Drift evaluation will need a point of reference on manual checking.
+
+Pros/Cons
+
+- 👍 Cover general cases of updates to the `NodePool` CRD or the hashing mechanism 
+- 👍 Handles cases where users change `NodePool` resources as Karpenter is updating
+- 👎 Cloud-provider would need to propagate on to the NodeClaim to adopt this method for static drift checking 
+- 👎👎 Negatively impact on performance for Karpenter until all NodeClaims match the `nodepool-hash-version`
```

Comment:
I'm thinking that in large clusters we will see an increases on the time in disruption and drift detection.

Reply:
I am not sure about disruption being slower but I think drift detection can be slower but it would be better to perform some test to see what the delta is. I honestly don't think we are talking about that much slower and disruption controller only trigger reconcile per nodeclaim update / create so worse case we would increase the number of workers on drift controller.

I could be wrong but unless we are changing how disruption controller works for drift, all it does is check `isDrifted` status. I don't see why this change would make that slower

Reply:
To me, we should almost completely remove this bullet. The biggest issue with this option (IMO) is the implementation complexity and the maintenance of having to constantly keep up with it